### PR TITLE
test: DatasourceControl

### PR DESCRIPTION
**Pull request by yardz**
### SUMMARY
Tests for DatasourceControl
- Moved to own folder
- Creating tests for the component

### TEST PLAN
No behavior changes and the interface for importing was maintained.
All tests must pass.

### CODE REVIEW

File: superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+test('Should render', () => {
+  render(<DatasourceControl {...defaultProps} />);
+  expect(screen.getByTestId('DatasourceControl')).toBeInTheDocument();
+});
+
+test('Should have elements', () => {
+  render(<DatasourceControl {...defaultProps} />);
+  expect(screen.getByText('channels')).toBeVisible();
+  expect(screen.getByTestId('datasource-menu-trigger')).toBeVisible(); // Error
```
Comment (michael-s-molina) [inline superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx:65]: What's this error comment?

File: superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+test('Should open a menu', () => {
+  render(<DatasourceControl {...defaultProps} />);
+
+  expect(screen.queryByText('Edit dataset')).not.toBeInTheDocument();
+  expect(screen.queryByText('Change dataset')).not.toBeInTheDocument();
+  expect(screen.queryByText('View in SQL Lab')).not.toBeInTheDocument();
+
+  userEvent.click(screen.getByTestId('datasource-menu-trigger'));
+
+  expect(screen.queryByText('Edit dataset')).toBeInTheDocument();
```
Comment (michael-s-molina) [inline superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx:77]:
```suggestion
  expect(screen.getByText('Edit dataset')).toBeInTheDocument();
```

File: superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  expect(screen.queryByText('Change dataset')).toBeInTheDocument();
```
Comment (michael-s-molina) [inline superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx:78]:
```suggestion
  expect(screen.getByText('Change dataset')).toBeInTheDocument();
```

File: superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  expect(screen.queryByText('View in SQL Lab')).toBeInTheDocument();
```
Comment (michael-s-molina) [inline superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx:79]:
```suggestion
  expect(screen.getByText('View in SQL Lab')).toBeInTheDocument();
```

File: superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  );
+
+  render(<DatasourceControl {...defaultProps} />, {
+    useRedux: true,
+  });
+  userEvent.click(screen.getByTestId('datasource-menu-trigger'));
+
+  await act(async () => {
+    userEvent.click(screen.getByText('Change dataset'));
+  });
```
Comment (michael-s-molina) [inline superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx:101]:
```suggestion
  userEvent.click(await screen.findByText('Change dataset'));
```
`findBy` methods are a combination of `getBy` queries and `waitFor`.

File: superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    async () => ({ json: { result: [] } } as any),
+  );
+  render(<DatasourceControl {...defaultProps} />, {
+    useRedux: true,
+  });
+  userEvent.click(screen.getByTestId('datasource-menu-trigger'));
+
+  await act(async () => {
+    userEvent.click(screen.getByText('Edit dataset'));
+  });
```
Comment (michael-s-molina) [inline superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx:120]:
```suggestion
  userEvent.click(await screen.findByText('Edit dataset'));
```

File: superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+  render(<DatasourceControl {...defaultProps} />, {
+    useRedux: true,
+  });
+  userEvent.click(screen.getByTestId('datasource-menu-trigger'));
+
+  expect(postFormSpy).toBeCalledTimes(0);
+
+  await act(async () => {
+    userEvent.click(screen.getByText('View in SQL Lab'));
+  });
```
Comment (michael-s-molina) [inline superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx:142]:
```suggestion
  userEvent.click(await screen.findByText('View in SQL Lab'));
```

File: superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx (hunk 5 above)
Comment (yardz) [reply superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx:101]: In this case the idea is to wait after the click. The element is already on the screen, the click must be within the `await act`.
this code `userEvent.click(await screen.findByText('Change dataset'));` produces a different result than expected.

File: superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx (hunk 6 above)
Comment (yardz) [reply superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx:120]: Same case here

File: superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx (hunk 7 above)
Comment (yardz) [reply superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx:142]: Same case...

**Review by geido — APPROVED** [review]
LGTM!

File: superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+import React from 'react';
+import { render, screen, act } from 'spec/helpers/testing-library';
+import userEvent from '@testing-library/user-event';
+import { SupersetClient } from '@superset-ui/core';
+import * as Utils from 'src/explore/exploreUtils';
+import DatasourceControl from '.';
+
+const SupersetClientGet = jest.spyOn(SupersetClient, 'get');
+
+let defaultProps: any;
```
Comment (michael-s-molina) [inline superset-frontend/src/explore/components/controls/DatasourceControl/DatasourceControl.test.tsx:29]: Remove `any` type.

**Review by michael-s-molina — APPROVED** [review]

**Review by rusackas — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**